spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -355,12 +372,15 @@ else if (output instanceof AbstractIntegrationMessageBuilder) {
 		else {
 			builder = this.getMessageBuilderFactory().withPayload(output);
 		}
-		if (this.shouldCopyRequestHeaders()) {
+		if (shouldCopyRequestHeaders()) {
```

Comment:
If (shouldCopyRequestHeaders() && !this.copyNoHeaders) {...

Reply:
I don't see reason in that `copyNoHeaders` `shouldCopyRequestHeaders` does the trick.

Reply:
Remember:
> if * {@link #shouldCopyRequestHeaders() shouldCopyRequestHeaaders} is true. 

?

So, independently of what you configure for the `headers`, but if `shouldCopyRequestHeaders()` returns `false`, they all are ignored.

Reply:
But what if a subclass overrides `shouldCopyRequestHeaders()` to return `true` and `*` is the only pattern?

Reply:
Ha! That was your suggestion to go pattern way 😄 
I was just suggested simple boolean flag - for symmetry of existing `shouldCopyRequestHeaders()`.

I don't know (yet) how to be...

Need to think, but I somehow don't like too much boolean flags...

Reply:
Perhaps we could call `shouldCopyRequestHeaders()` just once instead of on every message. It doesn't really make much sense to call it on every message anyway (unless a subclass used a `ThreadLocal`).